Feature/finish appeal aggregate

rhyslbw opened the pull request:
- reverses state check logic to target state when it makes sense
- rename `pledgeId` in pledge messages to `id`
- add appeal props to `AppealFulfilled` for richer projection capability
- add pledge props to pledge events for richer projection capability
- new error `AppealNotOpenToAcceptPledge`
- rename errors be more expressive
- tests reorganised to be more expressive

Todo:
- [x] Add `closed` concept
- [x] Rename first state of Pledge from `made` to `new` or similar. Pledges are still _made_ but that's an event, not a state.
- [x] Add `write off` pledge state and handling

rhyslbw committed:
Update to latest

- mgp deps
- AggregateRouter
- lowercase

rhyslbw committed:
Adds the concept of declining a pledge

rhyslbw committed:
improve appeal and pledge business rules

- reverses state check logic to target state when it makes sense
- rename `pledgeId` in pledge messages to `id`
- add appeal props to `AppealFulfilled` for richer projection capability
- add pledge props to pledge events for richer projection capability
- new error `AppealNotOpenToAcceptPledge`
- rename errors be more expressive
- tests reorganised to be more expressive

rhyslbw commented:
@DominikGuzei Please review for completeness, and I'll add anything else to the todo list

DominikGuzei commented on packages/domain/source/server/appeals/appeal.js:
```diff
@@ -42,31 +44,47 @@ Space.eventSourcing.Aggregate.extend(Donations, `Appeal`, {
+        }));
+      }
+    } else {
+      throw new Donations.AppealNotOpenForNewPledges();
```
Please throw early instead of wrapping everything into an `if` clause. So i would just turn this around -> if not has state open then throw immediately at the beginning.

rhyslbw replied:
k

DominikGuzei commented on packages/domain/source/server/appeals/appeal.js:
```diff
@@ -42,31 +44,47 @@ Space.eventSourcing.Aggregate.extend(Donations, `Appeal`, {
+    if (this.hasState(this.STATES.open)) {
+      this.record(new Donations.PledgeAccepted(_.extend({sourceId: this.getId()}, this._getPledgeById(command.id).toPlainObject())));
+    } else {
+      throw new Donations.AppealNotOpenToAcceptPledge();
```
same here

DominikGuzei commented on packages/domain/source/server/appeals/appeal.js:
```diff
@@ -42,31 +44,47 @@ Space.eventSourcing.Aggregate.extend(Donations, `Appeal`, {
+  },
+
+  _declinePledge(command) {
+    this.record(new Donations.PledgeDeclined(_.extend({sourceId: this.getId()}, this._getPledgeById(command.id).toPlainObject())));
```
would be more readable if you split it up into separate lines like:

```
this.record(new Donations.PledgeDeclined(_.extend({ sourceId: this.getId() }, 
  this._getPledgeById(command.id).toPlainObject()
)));
```

DominikGuzei commented on packages/domain/source/server/appeals/appeal.js:
```diff
@@ -42,31 +44,47 @@ Space.eventSourcing.Aggregate.extend(Donations, `Appeal`, {
 
   _fulfillPledge(command) {
-    this.record(new Donations.PledgeFulfilled(this._eventPropsFromCommand(command)));
+    this.record(new Donations.PledgeFulfilled(_.extend({sourceId: this.getId()}, this._getPledgeById(command.id).toPlainObject())));
```
same

DominikGuzei commented:
Looks great and quite complete already – your todo list covers the rest! If we find something is missing we can still add that in another PR :wink:

rhyslbw committed:
Throw errors early

Also some minor formatting improvements.

rhyslbw committed:
Add basic concept of closing appeal

Does not deal with the unfulfilled pledges yet

rhyslbw committed:
Pledge initial state now 'new'

- Protects fulfilled pledges from being accepted
- error renamed

rhyslbw committed:
Fix exception handling in Pledge entity

We need to protect against invariants before the event is saved, as
this is the state change definition. To achieve this while still
encapsulating the logic in the entity, I’ve separated the validation so
we can run the check from the Aggregate without changing the state

rhyslbw committed:
Add concept of reneging on pledge

rhyslbw committed:
Rename reneg concept as writing off

While the donor may _reneg_ on the fulfilment, it’s not the action the
admin takes to address this situation. The term would be better used to
describe the reason for writing off the pledge

rhyslbw commented:
Ok @DominikGuzei ready for review and merge. I've deferred the handling of unfinalised pledges at the time of _closing_ an appeal to a dedicated feature as It needs discussion, and it's not worth blocking us now.

DominikGuzei commented:
Looks great now! Yeah, let's not block too long – we should keep PRs small and focused. Merging now!
